readme.md
<p align="center">
  <img src="https://david-dm.org/nurdism/battleye/status.svg" alt="Dependencies Status">
  <img src="https://david-dm.org/nurdism/battleye/dev-status.svg" alt="Dev Dependencies Status">
  <img src="https://david-dm.org/nurdism/battleye/peer-status.svg" alt="Peer Dependencies Status">
  <a href="https://www.npmjs.com/package/battleye"><img src="https://img.shields.io/npm/dm/battleye.svg" alt="Downloads"></a>
  <a href="https://www.npmjs.com/package/battleye"><img src="https://img.shields.io/npm/v/battleye.svg" alt="Version"></a>
  <a href="https://www.npmjs.com/package/battleye"><img src="https://img.shields.io/npm/l/battleye.svg" alt="License"></a>
  <a href="https://discord.gg/Kzkd6V3" ><img src="https://discordapp.com/api/guilds/428366869993488401/widget.png" alt="Chat on discord"><a/>
</p>

# battleye

> Battleye rcon client built in nodejs.

## Example usage:
```js
import * as readline from 'readline'
import { readCfg, Socket } from 'battleye'

readCfg(process.cwd())
  .then(cfg => {
    console.log(cfg)

    // create socket
    const socket = new Socket({
      port: 2310,     // listen port
      ip: '0.0.0.0',  // listen ip
    })

    // create connection
    const connection = socket.connection({
      password: cfg.rconpassword,       // rcon password
      ip: cfg.rconip,                   // rcon ip
      port: parseInt(cfg.rconport, 10)  // rcon port
    }, {
      reconnect: true,            // reconnect on timeout
      reconnectTimeout: 500,      // how long (in ms) to try reconnect
      keepAlive: true,            // send keepAlive packet
      keepAliveInterval: 15000,   // keepAlive packet interval (in ms)
      timeout: true,              // timeout packets
      timeoutInterval: 1000,      // timeout packet check interval (in ms)
      timeoutThresholded: 5,      // packets to resend
      timeoutTime: 2000,          // interval to resend packet (in ms)
    })

    // create readline for command input
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    })

    socket.on('listening', (socket) => {
      const addr = socket.address()
      console.log(`Socket listening on ${typeof addr === 'string' ? addr : `${addr.address}:${addr.port}`}`)
    })

    socket.on('received', (resolved, packet, buffer, connection, info) => {
      console.log(`received: ${connection.ip}:${connection.port} => packet:`, packet)
    })

    socket.on('sent', (packet, buffer, bytes, connection) => {
      console.log(`sent: ${connection.ip}:${connection.port} => packet:`, packet)
    })

    socket.on('error', (err) => { console.error(`SOCKET ERROR:`, err) })

    connection.on('message', (message, packet) => {
      console.log(`message: ${connection.ip}:${connection.port} => message: ${message}`)
    })

    connection.on('command', (data, resolved, packet) => {
      console.log(`command: ${connection.ip}:${connection.port} => packet:`, packet)
    })

    connection.on('disconnected', (reason) => {
      console.warn(`disconnected from ${connection.ip}:${connection.port},`, reason)
    })

    connection.on('connected', () => {
      console.error(`connected to ${connection.ip}:${connection.port}`)
    })

    connection.on('debug', console.log)

    connection.on('error', (err) => {
      console.error(`CONNECTION ERROR:`, err)
    })

    rl.on('line', input => {
      connection
        .command(input)
        .then(response => {
          console.log(`response: ${connection.ip}:${connection.port} => ${response.command}\n${response.data}`)
        })
        .catch(console.error)

      console.log(`send: ${connection.ip}:${connection.port} => ${input}`)
    })
  })
  .catch(err => { console.error(`Error reading config:`, err) })
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.0.2](https://github.com/nurdism/battleye/compare/v1.0.1...v1.0.2) (2019-03-11)



## [1.0.1](https://github.com/nurdism/battleye/compare/v0.1.1...v1.0.1) (2019-03-10)



<a name="0.1.1"></a>
## 0.1.1 (2018-10-15)


src/connection.ts
import * as emitter from 'events'
import * as utils from './utils'

import { clearTimeout } from 'timers'
import { InvalidPassword, NoConnection, PacketOverflow, ServerDisconnect, ServerTimeout, UnknownCommand, UnknownPacketType } from './errors'
import { Packet, PacketDirection, PacketType } from './packet'
import { Socket } from './socket'

export interface IConnectionDetails {
  ip: string
  port: number
  password?: string
}

export interface IConnectionOptions {
  reconnect?: boolean
  reconnectTimeout?: number
  keepAlive?: boolean
  keepAliveInterval?: number
  timeout?: boolean
  timeoutInterval?: number
  timeoutThresholded?: number
  timeoutTime?: number
}

export interface IPacketPromise {
  bytes: number
  packet: Packet
  resolve (value?: IPacketResponse | PromiseLike<IPacketResponse>): void
  reject (reason?: Error): void
}

export interface IPacketResponse {
  command?: string
  data?: string
  connected?: boolean
  bytes: number
  sent: Packet
  received: Packet
  connection: Connection
}

export declare interface IConnection {
  on(event: 'message',      listener: (message: string, packet: Packet) => void): this;
  on(event: 'command',      listener: (data: string, resolved: boolean, packet: Packet) => void): this;

  on(event: 'connected',    listener: () => void): this;
  on(event: 'disconnected', listener: (reason: string | Error) => void): this;
  on(event: 'debug',        listener: (message: string) => void): this;
  on(event: 'error',        listener: (error: Error) => void): this;
}

/**
 * Connection to battleye rcon server
 *
 * @export
 * @class Connection
 * @extends {emitter}
 * @implements {IConnection}
 */
export class Connection extends emitter implements IConnection {
  private readonly socket: Socket
  private readonly options: IConnectionOptions
  private keepAlive: NodeJS.Timeout
  private timeout: NodeJS.Timeout
  private packets: [IPacketPromise, [IPacketPromise]]
  private multipart: [[Packet]]
  private readonly info: {
    id: string
    ip: string
    password: string
    port: number
    sent: number
    received: number
    sequence: number
    connected: boolean
  }

  /**
   * Creates an instance of Connection.
   *
   * @param {Socket} socket
   * @param {IConnectionDetails} details
   * @param {IConnectionOptions} [options={}]
   * @memberof Connection
   */
  constructor (socket: Socket, details: IConnectionDetails, options: IConnectionOptions = {}) {
    super()

    const { ip, port, password } = details

    this.info = {
      id: utils.hashAddress(ip, port),
      ip,
      password,
      port,
      sent: 0,
      received: 0,
      sequence: -1,
      connected: false,
    }

    this.socket = socket
    this.packets = [undefined, [undefined]]
    this.multipart = [undefined]

    this.options = {
      reconnect: true,
      reconnectTimeout: 500,
      keepAlive: true,
      keepAliveInterval: 15000,
      timeout: true,
      timeoutInterval: 1000,
      timeoutThresholded: 5,
      timeoutTime: 2000,
      ...options
    }
  }

  /**
   * connect to the connection
   *
   * @returns {Promise<IPacketResponse>}
   * @memberof Connection
   */
  public connect (): Promise<IPacketResponse> {
    if (!this.socket.listening) {
      throw new NoConnection()
    }

    if (this.connected) {
      this.disconnect()
    }

    this.setup()

    return this.send(new Packet(PacketType.Login, PacketDirection.Request, { password: this.info.password }))
  }

  /**
   * sends a command packet to connection
   *
   * @param {string} command
   * @param {boolean} [resolve=true]
   * @returns {Promise<IPacketResponse>}
   * @memberof Connection
   */
  public command (command: string, resolve: boolean = true): Promise<IPacketResponse> {
    if (!this.socket.listening || !this.connected) {
      throw new NoConnection()
    }

    return this.send(new Packet(PacketType.Command, PacketDirection.Request, { command }), resolve)
  }

  /**
   * sends a packet to connection
   *
   * @param {Packet} packet
   * @param {boolean} [resolve=true]
   * @returns {Promise<IPacketResponse>}
   * @memberof Connection
   */
  public send(packet: Packet, resolve: boolean = true): Promise<IPacketResponse> {
    return this.socket.send(this, packet, resolve)
  }

  /**
   * stores command Promise
   *
   * @param {IPacketPromise} store
   * @memberof Connection
   */
  public store (store: IPacketPromise): void {
    switch (store.packet.type) {
      case PacketType.Login:
        if (this.packets[0] !== undefined) { throw new PacketOverflow() }
        this.packets[0] = store
        break
      case PacketType.Command:
        if (this.packets[1][store.packet.sequence] !== undefined) { throw new PacketOverflow() }
        this.packets[1][store.packet.sequence] = store
        break
      default:
        throw new UnknownPacketType(store.packet.type)
    }
  }

  /**
   * resolves Promise from packet reply
   *
   * @param {Packet} packet
   * @returns {boolean}
   * @memberof Connection
   */
  public resolve (packet: Packet): boolean {
    let resolved = false;
    let store: IPacketPromise

    if (packet.direction === PacketDirection.Split) { // handle multipart packets

      if (this.multipart[packet.sequence] === undefined) {
        this.multipart[packet.sequence] = [undefined]
      }

      this.multipart[packet.sequence][packet.get('index')] = packet

      if ((packet.get('index') + 1) === packet.get('total')) {
        let buff = Buffer.alloc(0);
        for (const p of this.multipart[packet.sequence]) {
          const part = p.get('part')
          buff = Buffer.concat([buff, part], part.length + buff.length)
        }

        this.multipart.splice(packet.sequence, 1)
        return this.resolve(new Packet(PacketType.Command, PacketDirection.Reply, { data: buff.toString(), sequence: packet.sequence }))
      }

      return false
    }

    switch (packet.type) {
      case PacketType.Login:
        store = this.packets[0]

        this.info.connected = packet.get('login')

        if (store !== undefined) {
          if (this.connected) {
            store.resolve({
              connected: true,
              bytes: store.bytes,
              sent: store.packet,
              received: packet,
              connection: this
            })
          } else {
            store.reject(new InvalidPassword())
          }

          resolved = true
          this.packets[0] = undefined
        }

        if (this.connected) {
          this.emit('connected')
        } else {
          this.disconnect(new InvalidPassword())
        }
        break
      case PacketType.Command:
        store = this.packets[1][packet.sequence]

        if (store !== undefined) {
          if (packet.get('data') === 'Unknown command') {
            store.reject(new UnknownCommand(store.packet.get('command')))
          } else {
            store.resolve({
              command: store.packet.get('command'),
              data: packet.get('data'),
              bytes: store.bytes,
              sent: store.packet,
              received: packet,
              connection: this
            })
          }

          resolved = true
          this.packets[1].splice(packet.sequence, 1)
        }

        this.emit('command', packet.get('data'), resolved, packet)
        break
      case PacketType.Message:
        this.emit('message', packet.get('message'), packet)
        this
          .send(new Packet(PacketType.Message, PacketDirection.Reply, { sequence: packet.sequence }),false)
          .catch((e: Error)=> { this.emit('error', e) })
      break
      default:
        this.emit('error', new UnknownPacketType(packet.type))
    }

    return resolved
  }

  /**
   * returns weather or not connection is active
   *
   * @readonly
   * @type {boolean}
   * @memberof Connection
   */
  public get connected (): boolean {
    return this.info.connected
  }

  /**
   * returns connection id
   *
   * @readonly
   * @type {string}
   * @memberof Connection
   */
  public get id (): string {
    return this.info.id
  }

  /**
   * returns connection ip
   *
   * @readonly
   * @type {string}
   * @memberof Connection
   */
  public get ip (): string {
    return this.info.ip
  }

  /**
   * returns connection port
   *
   * @readonly
   * @type {number}
   * @memberof Connection
   */
  public get port (): number {
    return this.info.port
  }

  /**
   * kill connection
   *
   * @param {Error} reason
   * @memberof Connection
   */
  public kill(reason: Error): void {
    this.emit('error', reason)
    this.disconnect(reason)
  }

  /**
   * set's up timers for keep alive and timeouts
   *
   * @private
   * @memberof Connection
   */
  private setup (): void {
    const { keepAlive, keepAliveInterval, timeout, timeoutInterval, timeoutThresholded, timeoutTime } = this.options
    if (keepAlive && this.keepAlive === undefined) {
      this.keepAlive = setInterval(async () => {
        if (this.connected) {
          this
          .command('')
          .then(({ sent, received }: IPacketResponse) => {
            this.emit('debug', `ping: ${this.ip}:${this.port} ${(received.timestamp - sent.timestamp)}ms`)
          })
          .catch((e: Error) => { this.emit('error', e) })
        }
      }, keepAliveInterval)
    }

    if (timeout && this.timeout === undefined) {
      this.timeout = setInterval(async () => {
        const time = new Date().getTime()
        const check = (data: IPacketPromise) => {
          if (data !== undefined && data.packet instanceof Packet ) {
            if (time - data.packet.timestamp >= data.packet.sent * timeoutTime) {
              this
              .send(data.packet, false)
              .catch((e: Error) => this.emit('error', e))
            } else if (data.packet.sent >= timeoutThresholded) {
              this.disconnect(new ServerTimeout())
            }
          }
        }

        check(this.packets[0])
        for (const p of this.packets[1]) {
          check(p)
        }
      }, timeoutInterval)
    }
  }

  /**
   * returns connection packet sequence
   *
   * @readonly
   * @type {number}
   * @memberof Connection
   */
  public get sequence (): number {
    if (this.info.sequence > 255) { this.info.sequence = -1 }
    return this.info.sequence = this.info.sequence +1
  }

  /**
   * disconnects from connection
   *
   * @param {Error} [reason=new ServerDisconnect()]
   * @memberof Connection
   */
  private disconnect(reason: Error = new ServerDisconnect()): void {
    this.cleanup(reason)
    this.emit('disconnected', reason)
    const { reconnect, reconnectTimeout } = this.options
    if (reconnect && (reason instanceof ServerTimeout)) {
      setTimeout(() => {
        this
          .connect()
          .catch((e: Error) => {
            this.emit('error', e)
          })
      }, reconnectTimeout)
    }
  }

  /**
   * cleans up unresolved promises and resets connection
   *
   * @private
   * @param {Error} error
   * @memberof Connection
   */
  private cleanup (error: Error ): void {
    clearTimeout(this.timeout)
    clearTimeout(this.keepAlive)

    this.timeout = undefined
    this.keepAlive = undefined

    if (this.packets[0] !== undefined) {
      this.packets[0].reject(error)
    }

    for (const packet of this.packets[1]) {
      if (packet !== undefined) {
        packet.reject(error)
      }
    }

    this.info.sent = 0
    this.info.received = 0
    this.info.sequence = -1
    this.info.connected = false
    this.packets = [undefined, [undefined]]
    this.multipart = [undefined]
  }
}

